reco_utils/recommender/ncf/dataset.py

```diff
@@ -0,0 +1,301 @@
+                on=self.col_user,
+                how="left"
+            )
+            test_interact_status[self.col_item + '_negative'] = test_interact_status.apply(
+                    lambda row: row[self.col_item + '_negative'] - row[self.col_item + '_interacted_test'],
+                    axis=1,
+            )
+            test_ratings = pd.merge(
+                self.test,
+                test_interact_status[[self.col_user, self.col_item + '_negative']],
+                on=self.col_user,
+                how="left"
+            )
+
+            # sample n_neg_test negative samples for testing
+            try:
+                test_ratings[self.col_item + '_negative'] = test_ratings[self.col_item + '_negative'].apply(
+                    lambda x: random.sample(x, self.n_neg_test)
+                )
+
+            except:
+                min_num = min(map(len, list(test_ratings[self.col_item + '_negative'])))
+                warnings.warn("n_neg_test is larger than negative items set size! We will set n_neg as the smallest size: %d" % min_num)
+                test_ratings[self.col_item + '_negative'] = test_ratings[self.col_item + '_negative'].apply(
+                    lambda x: random.sample(x, min_num)
+                )
+
+            self.test_data = []
+
+            # generate test data
+            for row in test_ratings.itertuples():
+                self.test_users, self.test_items, self.test_ratings = [], [], []
+
+                self.test_users.append(int(getattr(row, self.col_user)))
+                self.test_items.append(int(getattr(row, self.col_item)))
+                self.test_ratings.append(float(getattr(row, self.col_rating)))
+
+                for i in getattr(row, self.col_item + '_negative'):
+                    self.test_users.append(int(getattr(row, self.col_user)))
+                    self.test_items.append(int(i))
+                    self.test_ratings.append(float(0))
+
+                self.test_data.append ( [
+                        [self.id2user[x] for x in self.test_users],
+                        [self.id2item[x] for x in self.test_items],
+                        self.test_ratings
+                    ]
+                )
+
+    def negative_sampling(self):
+        """ sample n_neg negative items per positive item
+            this function should be called every epoch
+        """
+        self.users, self.items, self.ratings = [], [], []
+
+        # sample n_neg negative samples for training
+        train_ratings = pd.merge(self.train, self.interact_status[[self.col_user, self.col_item + '_negative']],
+                                 on=self.col_user)
+
+        try:
+            train_ratings[self.col_item + '_negative'] = train_ratings[self.col_item + '_negative'].apply(
+                lambda x: random.sample(x, self.n_neg))
+        except:
+            raise ValueError
+            min_num = min(map(len, list(train_ratings[self.col_item + '_negative'])))
```

Comment:
This code is added in case the number of negative sampling `n_neg` is larger than the total number of negative items. If so, we will sample the negative items with smaller `n_neg` as the same as the smallest size of negative items.
Line 236 `raise ValueError` should be deleted, so the case is as the same as line 193. Thanks for your question, I will do it soon.

Reply:
cool, thanks